File: legacy-electron/lib/timer.js
// lib/timer.js — wall-clock pomodoro state machine (SPEC.md §5).
// No Electron, no timers of its own: the host drives it by calling tick() on
// whatever cadence it likes (main.js calls it every 250ms). Timing is stored
// as an epoch-ms `endsAt` rather than accumulated ticks, so it stays correct
// across sleep/wake — a tick() call after a long clock jump completes the
// phase in one step instead of drifting.
'use strict';

const { EventEmitter } = require('events');

const BREAK_PHASES = new Set(['shortBreak', 'longBreak']);

class Timer extends EventEmitter {
  constructor({ settingsGetter, now = Date.now } = {}) {
    super();
    this.getSettings = settingsGetter;
    this.now = now;

    this.phase = 'idle';
    this.running = false;
    this.ringing = false;
    this.endsAt = null;         // epoch ms while running; null while paused/idle
    this.remainingMs = 0;       // authoritative while not running
    this.totalMs = 0;           // duration locked in when the phase began
    this.cycleIndex = 0;
    this.completedToday = 0;
    this.completedTodayKey = null; // local calendar-day key completedToday is stamped with
    this.task = '';
    this.phaseStartedAt = null; // epoch ms the current phase first started running
    this.ringStartedAt = null;
  }

  // -- public API mirroring window.pomoppi (see SPEC.md §6) ---------------------

  start() {
    const nowTs = this.now();
    if (this.ringing) this._silenceRing(false);
    if (this.phase === 'idle') this._setupPhase('focus');
    if (this.running || this.remainingMs <= 0) return this.getState();
    this._beginRunning(nowTs);
    this.emit('change', this.getState());
    return this.getState();
  }

  pause() {
    const nowTs = this.now();
    if (this.ringing) this._silenceRing(false);
    if (!this.running) return this.getState();
    this.remainingMs = Math.max(0, this.endsAt - nowTs);
    this.endsAt = null;
    this.running = false;
    this.emit('change', this.getState());
    return this.getState();
  }

  reset() {
    const nowTs = this.now();
    if (this.ringing) this._silenceRing(false);
    if (this.phase !== 'idle' && this.phaseStartedAt != null) {
      const remaining = this._computeRemainingMs(nowTs);
      const actualMs = Math.max(0, this.totalMs - remaining);
      this._emitPhaseComplete({
        phase: this.phase,
        startedAt: this.phaseStartedAt,
        endedAt: nowTs,
        plannedMs: this.totalMs,
        actualMs,
        task: this.task,
        completed: false,
      });
    }
    this.phase = 'idle';
    this.running = false;
    this.endsAt = null;
    this.remainingMs = 0;
    this.totalMs = 0;
    this.phaseStartedAt = null;
    this.emit('change', this.getState());
    return this.getState();
  }

  // Ends the current phase early and moves straight on to the next one, either
  // direction: a break jumps into the focus session, a focus session is cut
  // short and hands over to the break. Idle is the only phase with nothing to
  // skip.
  //
  // The abandoned phase is always reported with `completed: false` and the time
  // actually spent, which is what makes lib/obsidian.js log it as a partial (an
  // ❌ line carrying `actualMs`, not the planned length, and excluded from the
  // day's focus total) whenever `logAborted` is on.
  //
  // A cut-short focus does NOT advance the cycle or the day's count: a long
  // break is earned by finishing sessions, not by skipping through them. So the
  // break that follows is always a short one, and `cycleIndex` stays put for
  // the next session to claim.
  //
  // Either direction bypasses autoStartFocus/autoStartBreaks — the user
  // explicitly asked to move on now.
  skip() {
    if (this.phase === 'idle') return this.getState();
    const nowTs = this.now();
    if (this.ringing) this._silenceRing(false);

    const skippedPhase = this.phase;
    const remaining = this._computeRemainingMs(nowTs);
    const actualMs = Math.max(0, this.totalMs - remaining);
    this._emitPhaseComplete({
      phase: skippedPhase,
      startedAt: this.phaseStartedAt != null ? this.phaseStartedAt : nowTs,
      endedAt: nowTs,
      plannedMs: this.totalMs,
      actualMs,
      task: this.task,
      completed: false,
    });

    if (BREAK_PHASES.has(skippedPhase)) {
      this._setupPhase('focus');
    } else {
      // The task belonged to the session just abandoned; it went out with the
      // phaseComplete above, same as when a focus session runs to the end.
      this.task = '';
      this._setupPhase('shortBreak');
    }
    this._beginRunning(nowTs);
    this.emit('change', this.getState());
    return this.getState();
  }

  dismissRing() {
    if (this.ringing) this._silenceRing(true);
    return this.getState();
  }

  setTask(task) {
    this.task = typeof task === 'string' ? task : '';
    this.emit('change', this.getState());
    return this.getState();
  }

  getTask() {
    return this.task;
  }

  // Call on whatever cadence the host wants (main.js: every 250ms). Detects
  // phase completion and ring timeout, then always emits 'tick'.
  tick() {
    const nowTs = this.now();
    if (this.ringing) {
      const ringMs = Math.max(0, Number(this.getSettings().ringSeconds) || 0) * 1000;
      if (nowTs - this.ringStartedAt >= ringMs) this._silenceRing(false);
    }
    if (this.running && this.endsAt != null && nowTs >= this.endsAt) {
      this._completePhase(nowTs);
    }
    const state = this.getState();
    this.emit('tick', state);
    return state;
  }

  // While idle there is no phase counting down, but the widget still has a
  // clock to fill: report the focus length that pressing start would use, so a
  // timer that has never run reads 25:00 instead of 00:00. Derived on read
  // rather than stored, so changing focusMinutes while idle moves the preview
  // with no extra plumbing, and start() locks in the same number.
  getState() {
    this._rolloverIfNeeded(this.now());
    const idle = this.phase === 'idle';
    const totalMs = idle ? this._durationMsFor('focus') : this.totalMs;
    return {
      phase: this.phase,
      running: this.running,
      ringing: this.ringing,
      remainingMs: idle ? totalMs : Math.max(0, this._computeRemainingMs(this.now())),
      totalMs,
      cycleIndex: this.cycleIndex,
      completedToday: this.completedToday,
      task: this.task,
    };
  }

  // -- internals ---------------------------------------------------------

  // completedToday is stamped with the local calendar day it belongs to; a
  // widget left running overnight rolls it back to 0 rather than carrying
  // yesterday's total into today.
  _dateKey(nowTs) {
    const d = new Date(nowTs);
    return `${d.getFullYear()}-${d.getMonth()}-${d.getDate()}`;
  }

  _rolloverIfNeeded(nowTs) {
    const key = this._dateKey(nowTs);
    if (this.completedTodayKey !== null && this.completedTodayKey !== key) {
      this.completedToday = 0;
    }
    this.completedTodayKey = key;
  }

  _computeRemainingMs(nowTs) {
    if (this.running && this.endsAt != null) return Math.max(0, this.endsAt - nowTs);
    return Math.max(0, this.remainingMs);
  }

  _durationMsFor(phase) {
    const s = this.getSettings();
    if (phase === 'focus') return s.focusMinutes * 60000;
    if (phase === 'shortBreak') return s.shortBreakMinutes * 60000;
    if (phase === 'longBreak') return s.longBreakMinutes * 60000;
    return 0;
  }

  _setupPhase(phase) {
    this.phase = phase;
    this.totalMs = this._durationMsFor(phase);
    this.remainingMs = this.totalMs;
    this.endsAt = null;
    this.running = false;
    this.phaseStartedAt = null;
  }

  _beginRunning(nowTs) {
    if (this.phaseStartedAt == null) this.phaseStartedAt = nowTs;
    this.endsAt = nowTs + this.remainingMs;
    this.running = true;
  }

  _silenceRing(emitChange) {
    this.ringing = false;
    this.ringStartedAt = null;
    if (emitChange) this.emit('change', this.getState());
  }

  _emitPhaseComplete(payload) {
    this.emit('phaseComplete', payload);
  }

  _completePhase(nowTs) {
    const finishedPhase = this.phase;
    const startedAt = this.phaseStartedAt != null ? this.phaseStartedAt : nowTs;
    const plannedMs = this.totalMs;
    const task = this.task;

    this.running = false;
    this.endsAt = null;
    this.remainingMs = 0;

    this._emitPhaseComplete({
      phase: finishedPhase,
      startedAt,
      endedAt: nowTs,
      plannedMs,
      actualMs: plannedMs, // ran to completion, so actual == planned
      task,
      completed: true,
    });

    const settings = this.getSettings();
    let nextPhase;
    if (finishedPhase === 'focus') {
      this._rolloverIfNeeded(nowTs);
      this.completedToday += 1;
      this.cycleIndex += 1;
      if (this.cycleIndex >= settings.longBreakEvery) {
        nextPhase = 'longBreak';
        this.cycleIndex = 0;
      } else {
        nextPhase = 'shortBreak';
      }
      this.task = '';
    } else {
      nextPhase = 'focus';
    }

    this._setupPhase(nextPhase);
    this.ringing = true;
    this.ringStartedAt = nowTs;

    const autoStart = nextPhase === 'focus' ? settings.autoStartFocus : settings.autoStartBreaks;
    if (autoStart) this._beginRunning(nowTs);

    this.emit('change', this.getState());
  }
}

module.exports = Timer;

File: native/Sources/PomoppiCore/Timer.swift
// Timer.swift — wall-clock pomodoro state machine.
// Port of lib/timer.js (SPEC.md §5): timing is stored as an absolute `endsAt`
// rather than accumulated ticks, so it stays correct across sleep/wake — a
// tick() call after a long clock jump completes the phase in one step
// instead of drifting. Named `PomodoroTimer`, not `Timer`, to avoid colliding
// with Foundation.Timer.
import Foundation

public enum Phase: String, Equatable {
    case idle, focus, shortBreak, longBreak
}

public struct TimerState: Equatable {
    public let phase: Phase
    public let running: Bool
    public let ringing: Bool
    public let remainingMs: Double
    public let totalMs: Double
    public let cycleIndex: Int
    public let completedToday: Int
    public let task: String
}

public struct PhaseCompleteEvent {
    public let phase: Phase
    public let startedAt: Date
    public let endedAt: Date
    public let plannedMs: Double
    public let actualMs: Double
    public let task: String
    public let completed: Bool
}

// The slice of settings the timer needs, decoupled from the full Settings
// model so this file (and its tests) don't depend on PomoppiCore's disk I/O.
public struct TimerSettingsSnapshot {
    public var focusMinutes: Double
    public var shortBreakMinutes: Double
    public var longBreakMinutes: Double
    public var longBreakEvery: Int
    public var autoStartBreaks: Bool
    public var autoStartFocus: Bool
    public var ringSeconds: Double

    public init(
        focusMinutes: Double,
        shortBreakMinutes: Double,
        longBreakMinutes: Double,
        longBreakEvery: Int,
        autoStartBreaks: Bool,
        autoStartFocus: Bool,
        ringSeconds: Double
    ) {
        self.focusMinutes = focusMinutes
        self.shortBreakMinutes = shortBreakMinutes
        self.longBreakMinutes = longBreakMinutes
        self.longBreakEvery = longBreakEvery
        self.autoStartBreaks = autoStartBreaks
        self.autoStartFocus = autoStartFocus
        self.ringSeconds = ringSeconds
    }
}

private let breakPhases: Set<Phase> = [.shortBreak, .longBreak]

public final class PomodoroTimer {
    public var onChange: ((TimerState) -> Void)?
    public var onTick: ((TimerState) -> Void)?
    public var onPhaseComplete: ((PhaseCompleteEvent) -> Void)?

    private let getSettings: () -> TimerSettingsSnapshot
    private let now: () -> Date

    private var phase: Phase = .idle
    private var running = false
    private var ringing = false
    private var endsAt: Date?          // set while running; nil while paused/idle
    private var remainingMs: Double = 0 // authoritative while not running
    private var totalMs: Double = 0     // duration locked in when the phase began
    private var cycleIndex = 0
    private var completedToday = 0
    private var completedTodayKey: String?
    private var task: String = ""
    private var phaseStartedAt: Date?
    private var ringStartedAt: Date?

    public init(settingsGetter: @escaping () -> TimerSettingsSnapshot, now: @escaping () -> Date = Date.init) {
        self.getSettings = settingsGetter
        self.now = now
    }

    // -- public API mirroring lib/timer.js -----------------------------

    @discardableResult
    public func start() -> TimerState {
        let nowTs = now()
        if ringing { silenceRing(emitChange: false) }
        if phase == .idle { setupPhase(.focus) }
        if running || remainingMs <= 0 { return getState() }
        beginRunning(nowTs)
        onChange?(getState())
        return getState()
    }

    @discardableResult
    public func pause() -> TimerState {
        let nowTs = now()
        if ringing { silenceRing(emitChange: false) }
        guard running else { return getState() }
        remainingMs = max(0, computeRemainingMs(nowTs))
        endsAt = nil
        running = false
        onChange?(getState())
        return getState()
    }

    @discardableResult
    public func reset() -> TimerState {
        let nowTs = now()
        if ringing { silenceRing(emitChange: false) }
        if phase != .idle, let startedAt = phaseStartedAt {
            let remaining = computeRemainingMs(nowTs)
            let actualMs = max(0, totalMs - remaining)
            emitPhaseComplete(PhaseCompleteEvent(
                phase: phase, startedAt: startedAt, endedAt: nowTs,
                plannedMs: totalMs, actualMs: actualMs, task: task, completed: false))
        }
        phase = .idle
        running = false
        endsAt = nil
        remainingMs = 0
        totalMs = 0
        phaseStartedAt = nil
        onChange?(getState())
        return getState()
    }

    // Ends the current phase early and moves straight on to the next one,
    // either direction. A cut-short focus earns no cycle/day credit — a long
    // break is earned by finishing sessions, not by skipping through them —
    // and bypasses autoStartFocus/autoStartBreaks.
    @discardableResult
    public func skip() -> TimerState {
        guard phase != .idle else { return getState() }
        let nowTs = now()
        if ringing { silenceRing(emitChange: false) }

        let skippedPhase = phase
        let remaining = computeRemainingMs(nowTs)
        let actualMs = max(0, totalMs - remaining)
        emitPhaseComplete(PhaseCompleteEvent(
            phase: skippedPhase, startedAt: phaseStartedAt ?? nowTs, endedAt: nowTs,
            plannedMs: totalMs, actualMs: actualMs, task: task, completed: false))

        if breakPhases.contains(skippedPhase) {
            setupPhase(.focus)
        } else {
            task = ""
            setupPhase(.shortBreak)
        }
        beginRunning(nowTs)
        onChange?(getState())
        return getState()
    }

    @discardableResult
    public func dismissRing() -> TimerState {
        if ringing { silenceRing(emitChange: true) }
        return getState()
    }

    @discardableResult
    public func setTask(_ newTask: String) -> TimerState {
        task = newTask
        onChange?(getState())
        return getState()
    }

    public func getTask() -> String { task }

    // Call on whatever cadence the host wants. Detects phase completion and
    // ring timeout, then always fires onTick.
    @discardableResult
    public func tick() -> TimerState {
        let nowTs = now()
        if ringing, let ringStartedAt {
            let ringMs = max(0, getSettings().ringSeconds) * 1000
            if nowTs.timeIntervalSince(ringStartedAt) * 1000 >= ringMs {
                silenceRing(emitChange: false)
            }
        }
        if running, let endsAt, nowTs >= endsAt {
            completePhase(nowTs)
        }
        let state = getState()
        onTick?(state)
        return state
    }

    // While idle there is no phase counting down, but the widget still has a
    // clock to fill: report the focus length that pressing start would use,
    // derived on read rather than stored, so changing focusMinutes while idle
    // moves the preview with no extra plumbing.
    public func getState() -> TimerState {
        rolloverIfNeeded(now())
        let idle = phase == .idle
        let stateTotalMs = idle ? durationMs(for: .focus) : totalMs
        let stateRemainingMs = idle ? stateTotalMs : max(0, computeRemainingMs(now()))
        return TimerState(
            phase: phase, running: running, ringing: ringing,
            remainingMs: stateRemainingMs, totalMs: stateTotalMs,
            cycleIndex: cycleIndex, completedToday: completedToday, task: task)
    }

    // -- internals -------------------------------------------------------

    // completedToday is stamped with the local calendar day it belongs to; a
    // widget left running overnight rolls it back to 0 rather than carrying
    // yesterday's total into today.
    private func dateKey(_ nowTs: Date) -> String {
        let comps = Calendar.current.dateComponents([.year, .month, .day], from: nowTs)
        return "\(comps.year ?? 0)-\(comps.month ?? 0)-\(comps.day ?? 0)"
    }

    private func rolloverIfNeeded(_ nowTs: Date) {
        let key = dateKey(nowTs)
        if let currentKey = completedTodayKey, currentKey != key {
            completedToday = 0
        }
        completedTodayKey = key
    }

    private func computeRemainingMs(_ nowTs: Date) -> Double {
        if running, let endsAt {
            return max(0, endsAt.timeIntervalSince(nowTs) * 1000)
        }
        return max(0, remainingMs)
    }

    private func durationMs(for phase: Phase) -> Double {
        let s = getSettings()
        switch phase {
        case .focus: return s.focusMinutes * 60000
        case .shortBreak: return s.shortBreakMinutes * 60000
        case .longBreak: return s.longBreakMinutes * 60000
        case .idle: return 0
        }
    }

    private func setupPhase(_ newPhase: Phase) {
        phase = newPhase
        totalMs = durationMs(for: newPhase)
        remainingMs = totalMs
        endsAt = nil
        running = false
        phaseStartedAt = nil
    }

    private func beginRunning(_ nowTs: Date) {
        if phaseStartedAt == nil { phaseStartedAt = nowTs }
        endsAt = nowTs.addingTimeInterval(remainingMs / 1000)
        running = true
    }

    private func silenceRing(emitChange: Bool) {
        ringing = false
        ringStartedAt = nil
        if emitChange { onChange?(getState()) }
    }

    private func emitPhaseComplete(_ event: PhaseCompleteEvent) {
        onPhaseComplete?(event)
    }

    private func completePhase(_ nowTs: Date) {
        let finishedPhase = phase
        let startedAt = phaseStartedAt ?? nowTs
        let plannedMs = totalMs
        let finishedTask = task

        running = false
        endsAt = nil
        remainingMs = 0

        emitPhaseComplete(PhaseCompleteEvent(
            phase: finishedPhase, startedAt: startedAt, endedAt: nowTs,
            plannedMs: plannedMs, actualMs: plannedMs, task: finishedTask, completed: true))

        let settings = getSettings()
        let nextPhase: Phase
        if finishedPhase == .focus {
            rolloverIfNeeded(nowTs)
            completedToday += 1
            cycleIndex += 1
            if cycleIndex >= settings.longBreakEvery {
                nextPhase = .longBreak
                cycleIndex = 0
            } else {
                nextPhase = .shortBreak
            }
            task = ""
        } else {
            nextPhase = .focus
        }

        setupPhase(nextPhase)
        ringing = true
        ringStartedAt = nowTs

        let autoStart = nextPhase == .focus ? settings.autoStartFocus : settings.autoStartBreaks
        if autoStart { beginRunning(nowTs) }

        onChange?(getState())
    }
}
